### PR TITLE
Configuration for report file location

### PR DESCRIPTION
**Pull request by nderwin**
* added runtime configuration to enable / disable building the reports
* added source (.jrxml) and destination (.jasper) paths for the reports
* ignore another NetBeans specific file for git

Signed-off-by:Nathan Erwin <nathan.d.erwin@gmail.com>

### CODE REVIEW

**Comment by melloware** [comment]
in `\runtime\src\main\resources\META-INF\quarkus-extension.yaml` make sure to add the `config` line below now that we have configs.

```yml
name: JasperReports
description: JasperReports is an open source reporting library that can be embedded into any Java application.
artifact: ${project.groupId}:${project.artifactId}:${project.version}
metadata:
  short-name: "jasperreports"
  type: "test"
  keywords:
    - "reports"
    - "reporting"
    - "pdf"
    - "excel"
  categories:
    - "miscellaneous"
  config:
    - "quarkus.jasperreports."
  status: "preview"
  guide: "https://docs.quarkiverse.io/quarkus-jasperreports/dev/index.html"
  icon-url: "https://raw.githubusercontent.com/quarkiverse/quarkus-jasperreports/main/docs/modules/ROOT/assets/images/jasperreports.svg"
```

**Comment by nderwin** [comment]
> in `\runtime\src\main\resources\META-INF\quarkus-extension.yaml` make sure to add the `config` line below now that we have configs.

Done.  Now, to figure out _why_ it's not adding the config to the generated docs.  The "Writing You Own Extension" guide makes it sound like you don't have to do anything but add the config class(es).

**Comment by nderwin** [comment]
https://github.com/quarkiverse/quarkus-jasperreports/pull/99#issuecomment-2398122459

@gastaldi I do the same, use a Maven plugin, but the purpose of this is to roll that into the deployment, though I'm not sure I'm producing the correct build item as I'm not getting a .jasper file to show up in the target directory or the .jar file.

**Comment by gastaldi** [comment]
Perhaps we need a processor producing a `CompiledReportBuildItem` with the path to the compiled jasper? Just thinking out loud

**Comment by nderwin** [comment]
@gastaldi Ok, I'm stuck.  I tried writing the `CompiledReportBuildItem`, but then a build step has to consume it, and then do what with it?  I just can't get it to output a `.jasper` file into the target directory someplace, or the `.jar`, and it's probably something really simple that I missed.  Right now, I'm using a `GeneratedClassBuildItem` like what is being used for the native output.

**Comment by melloware** [comment]
@nderwin i think I can help with this. I will look tomorrow.

**Comment by melloware** [comment]
This is just pseudocode i didn't test it but to save to the output target directory the .jasper file.

```java
@BuildStep
    void compileReports(ReportConfig config, List<ReportFileBuildItem> reportFiles,
            BuildProducer<GeneratedClassBuildItem> compiledReportProducer,
			OutputTargetBuildItem outputTarget) {

        Log.warnf("Found %s report(s) to compile? %s", reportFiles.size(), config.build().enable());
        if (config.build().enable()) {
            // TODO - only compile if the report file has changed
            for (ReportFileBuildItem item : reportFiles) {
                String outputFile = item.getFileName().replace("." + ReportFileBuildItem.EXT_REPORT,
                        "." + ReportFileBuildItem.EXT_COMPILED);
                String outputFilePath = outputTarget.getOutputDirectory().get(Path.of(config.build().destination().toString(), outputFile)).toString();
                Log.warnf("Compiling %s into %s", item.getFileName(), outputFilePath);

                try (InputStream inputStream = JRLoader.getLocationInputStream(item.getFileName())) {
                    ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
                    JasperCompileManager.compileReportToStream(inputStream, outputStream);
					// might have to reset or reload the input stream if its a bytearrayStream you can call reset
					JasperDesign jasperDesign = JRXmlLoader.load(inputStream);
					JasperCompileManager.compileReportToFile(jasperDesign, outputFilePath);

                    Log.warnf("Compiled size is %s", outputStream.size());
                    compiledReportProducer
                            .produce(new GeneratedClassBuildItem(true,
                                    outputFilePath, // TODO: this is supposed to be the class name ??
                                    outputStream.toByteArray()));
                } catch (JRException ex) {
                } catch (IOException ex) {
                }
            }
        } else {
            Log.debug("Automatic report compilation disabled");
        }
    }
```

It uses the `OutpputTargetBuildItem` to get the target directory and then you have to write the Jasper compiled file to that directory.

**Comment by nderwin** [comment]
@melloware Yes, now I have generated .jasper files.  Needs some more cleanup, but this should be ready late today, and then it can be iterated and improved upon.

File: deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java [hunk 1]
```diff
@@ -413,12 +419,62 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
...
+            BuildProducer<GeneratedClassBuildItem> compiledReportProducer, OutputTargetBuildItem outputTarget) {
+
+        Log.warnf("Found %s report(s) to compile? %s", reportFiles.size(), config.build().enable());
+        if (config.build().enable()) {
+            // make sure the destination path exists before we try to write files to it
+            Path outputFilePath = Path.of(outputTarget.getOutputDirectory().toString(), "classes",
+                    config.build().destination().toString());
+            if (!outputFilePath.toFile().exists()) {
+                outputFilePath.toFile().mkdirs();
+            }
```
Comment (gastaldi) [inline deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java:449]:
```suggestion
            if (!Files.exists(outputFilePath.getParent())) {
                Files.createDirectories(outputFilePath.getParent());
            }
```

File: deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java [hunk 2]
```diff
@@ -413,12 +419,69 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
...
+
+        if (config.build().enable()) {
+            Log.debugf("Found %s report(s) to compile?", reportFiles.size());
+            try {
+                // make sure the destination path exists before we try to write files to it
+                Path outputFilePath = Path.of(outputTarget.getOutputDirectory().toString(), "classes",
+                        config.build().destination().get().toString());
+
+                if (!Files.exists(outputFilePath)) {
+                    Files.createDirectories(outputFilePath);
```
Comment (gastaldi) [inline deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java:450]: You need to call `outputFilePath.getParent()` if I am not mistaken

Comment (nderwin) [reply deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java:448]: I did try that at first, but it didn't work right.  This one puts the output directory where I wanted it to be.

File: deployment/src/main/java/io/quarkiverse/jasperreports/deployment/ReportConfig.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        /**
+         * Enable building all report files.
+         */
+        @WithDefault("true")
+        boolean enable();
+
+        /**
+         * The path where all source .jrxml files are located.
+         */
+        @WithDefault("/src/main/jasperreports")
```
Comment (gastaldi) [inline deployment/src/main/java/io/quarkiverse/jasperreports/deployment/ReportConfig.java:31]: Shouldn't it be: 

```suggestion
        @WithDefault("src/main/jasperreports")
```

Comment (nderwin) [reply deployment/src/main/java/io/quarkiverse/jasperreports/deployment/ReportConfig.java]: Yes!  Fixed.

**Comment by nderwin** [comment]
@melloware This is so close...  just barfing on the native tests if you could take a look?

**Comment by melloware** [comment]
Have you merged all my latest changes from main?

**Comment by melloware** [comment]
Oh I bet I know is it trying to compile on build?  You can't compile in native mode so you have to add a LauncHMode check to your build step so it doesn't happen in Native mode.

**Comment by nderwin** [comment]
Yup, has the latest;  I'm not as familiar with native mode (haven't needed to use it), but I thought we'd be able to use the output of the compile step to feed into whatever native mode needs?

**Comment by melloware** [comment]
I will take a closer look tomorrow.

**Comment by melloware** [comment]
Found the issue. You need to put the code back the way I had it because it runs the integration tests against the native build and that absolutely cannot compile.

**Comment by gastaldi** [comment]
> I will take a closer look tomorrow.

Wow, your days are really short :)

File: integration-tests/src/main/java/io/quarkiverse/jasperreports/it/AbstractJasperResource.java [hunk 4]
```diff
@@ -4,24 +4,28 @@
...
-        if (isRunningInContainer()) {
-            result = (JasperReport) JRLoader.loadObject(JRLoader.getLocationInputStream(jasperFile + ".jasper"));
-        } else {
-            result = compile(jasperFile);
-        }
+        // TODO - Do we need this check anymore?
+        //        if (isRunningInContainer()) {
+        //            result = (JasperReport) JRLoader.loadObject(JRLoader.getLocationInputStream(jasperFile + ".jasper"));
+        //        } else {
+        result = compile(jasperFile);
```
[the remark this answers is not in the document]
Comment (nderwin) [reply integration-tests/src/main/java/io/quarkiverse/jasperreports/it/AbstractJasperResource.java:27]: `compile` doesn't actually compile any more, it basically does what line 25 is doing - loading the previously compiled `.jasper` file.

**Comment by melloware** [comment]
I am seeing errors in the log saying "src/main/resources" is invalid. Like it's not actually compiling them?

**Comment by melloware** [comment]
@nderwin i am going to merge this and work on a fix. Its easier for me to figure it out working locally.

**Comment by nderwin** [comment]
@melloware Ok, let me know if there's anything I can do to help.

**Pull request merged**